autoload/vista/parser/lsp.vim

```diff
@@ -41,80 +43,79 @@ function! s:IsFileUri(uri) abort
-      endfor
-    endif
-  endif
+function! s:LspToLocalSymbol(sym, range)
+  return {
+    \ 'lnum': a:range.start.line + 1,
+    \ 'col': a:range.start.character + 1,
+    \ 'kind': s:Kind2Symbol(a:sym.kind),
+    \ 'text': a:sym.name,
+    \ }
 endfunction
 
-function! vista#parser#lsp#CocSymbols(symbol, container) abort
-  if vista#ShouldIgnore(a:symbol.kind)
-    return
-  endif
+function! s:LocalToRawSymbol(sym)
+  return {
+    \ 'line': a:sym.lnum,
+    \ 'kind': a:sym.kind,
+    \ 'name': a:sym.text,
+    \ }
+endfunction
 
-  let raw = { 'line': a:symbol.lnum, 'kind': a:symbol.kind, 'name': a:symbol.text }
-  call add(g:vista.raw, raw)
+function! s:IsDocumentSymbol(sym)
+  return has_key(a:sym, 'selectionRange')
+endfunction
 
-  if a:symbol.kind ==? 'Method' || a:symbol.kind ==? 'Function'
-    call add(g:vista.functions, a:symbol)
-  endif
+function! s:ParseSymbolInfoList(symbols) abort
+  let filtered = filter(a:symbols, 's:IsFileUri(v:val.location.uri)')
+  return map(filtered, 's:LspToLocalSymbol(v:val, v:val.location.range)')
+endfunction
 
-  call add(a:container, {
-        \ 'lnum': a:symbol.lnum,
-        \ 'col': a:symbol.col,
-        \ 'text': a:symbol.text,
-        \ 'kind': a:symbol.kind,
-        \ 'level': a:symbol.level
-        \ })
+function! s:ParseDocumentSymbolsRec(outlist, symbols, level) abort
+  for lspsym in a:symbols
+    let sym = s:LspToLocalSymbol(lspsym, lspsym.selectionRange)
+    let sym.level = a:level
+    call add(a:outlist, sym)
+    if has_key(lspsym, 'children')
+      call s:ParseDocumentSymbolsRec(a:outlist, lspsym.children, a:level + 1)
+    endif
+  endfor
+  return a:outlist
 endfunction
 
-" https://microsoft.github.io/language-server-protocol/specification#textDocument_documentSymbol
-function! vista#parser#lsp#ExtractSymbol(symbol, container) abort
-  let symbol = a:symbol
+function! s:GroupSymbolsByKind(symbols) abort
+  let groups = {}
+  for sym in a:symbols
+    if has_key(groups, sym.kind)
+      call add(groups[sym.kind], sym)
+    else
+      let groups[sym.kind] = [ sym ]
+    endif
+  endfor
+  return groups
+endfunction
 
-  if vista#ShouldIgnore(symbol.kind)
-    return
+function! vista#parser#lsp#ParseDocumentSymbolPayload(resp) abort
+  if s:IsDocumentSymbol(a:resp[0])
+    let symbols = s:ParseDocumentSymbolsRec([], a:resp, 0)
+    return vista#parser#lsp#DispatchDocumentSymbols(symbols)
+  else
+    let symbols = s:ParseSymbolInfoList(a:resp)
+    call s:FilterDocumentSymbols(symbols)
+    return s:GroupSymbolsByKind(symbols)
   endif
+endfunction
 
-  if symbol.kind ==? 'Method' || symbol.kind ==? 'Function'
-    call add(g:vista.functions, symbol)
+function! s:FilterDocumentSymbols(symbols) abort
+  let symlist = a:symbols
+  if exists('g:vista_ignore_kinds')
+    call filter(symlist, 'index(g:vista_ignore_kinds, v:val) < 0')
   endif
+  let g:vista.functions =
+    \ filter(copy(symlist), 'v:val.kind ==? "Method" || v:val.kind ==? "Function"')
```

Comment:
Not sure which one is better, if you put this filter logic into `s:GroupSymbolsByKind()`, then you have only one-for loop, these two `fitler` can be saved.